cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceFactory.java

```diff
@@ -52,111 +53,154 @@
   }
 
-  static SparkBatchSourceFactory create(InputFormatProvider inputFormatProvider) {
-    return new SparkBatchSourceFactory(null, inputFormatProvider, null);
+  private final Map<String, StreamBatchReadable> streamBatchReadables;
+  private final Map<String, InputFormatProvider> inputFormatProviders;
+  private final Map<String, DatasetInfo> datasetInfos;
+  private final Map<String, Set<String>> sourceInputs;
+
+  SparkBatchSourceFactory() {
+    this.streamBatchReadables = new HashMap<>();
+    this.inputFormatProviders = new HashMap<>();
+    this.datasetInfos = new HashMap<>();
+    this.sourceInputs = new HashMap<>();
   }
 
-  static SparkBatchSourceFactory create(String datasetName, Map<String, String> datasetArgs,
-                                        @Nullable List<Split> splits) {
-    return new SparkBatchSourceFactory(null, null, new DatasetInfo(datasetName, datasetArgs, splits));
+  private SparkBatchSourceFactory(Map<String, StreamBatchReadable> streamBatchReadables,
+                                  Map<String, InputFormatProvider> inputFormatProviders,
+                                  Map<String, DatasetInfo> datasetInfos,
+                                  Map<String, Set<String>> sourceInputs) {
+    this.streamBatchReadables = streamBatchReadables;
+    this.inputFormatProviders = inputFormatProviders;
+    this.datasetInfos = datasetInfos;
+    this.sourceInputs = sourceInputs;
   }
 
-  static SparkBatchSourceFactory create(Input input) {
+  public void addInput(String stageName, Input input) {
     if (input instanceof Input.DatasetInput) {
       // Note if input format provider is trackable then it comes in as DatasetInput
       Input.DatasetInput datasetInput = (Input.DatasetInput) input;
-      return create(datasetInput.getName(), datasetInput.getArguments(), datasetInput.getSplits());
-    } else if (input instanceof Input.StreamInput) {
-      Input.StreamInput streamInput = (Input.StreamInput) input;
-      return create(streamInput.getStreamBatchReadable());
+      addInput(stageName, datasetInput.getName(), datasetInput.getAlias(), datasetInput.getArguments());
     } else if (input instanceof Input.InputFormatProviderInput) {
       Input.InputFormatProviderInput ifpInput = (Input.InputFormatProviderInput) input;
-      return new SparkBatchSourceFactory(null, ifpInput.getInputFormatProvider(), null);
+      addInput(stageName, ifpInput.getAlias(),
+               new BasicInputFormatProvider(ifpInput.getInputFormatProvider().getInputFormatClassName(),
+                                            ifpInput.getInputFormatProvider().getInputFormatConfiguration()));
+    } else if (input instanceof Input.StreamInput) {
+      Input.StreamInput streamInput = (Input.StreamInput) input;
+      addInput(stageName, streamInput.getAlias(), streamInput.getStreamBatchReadable());
     }
-    throw new IllegalArgumentException("Unknown input format type: " + input.getClass().getCanonicalName());
```

Comment:
In the previous logic, we used a byte to represent the input type - so the inputStream starts with a byte that says what the input type is. But now we have three maps holding that info (one for Streams, one for Datasets, one for InputFormatProviders) and so we can't have any other type. And we have only one public method ```addInput``` which takes in an Input and that can be of only three types - ```Input.DatasetInput```, ```Input.StreamInput```, ```Input.InputFormatProviderInput```. So we are assured that we won't get an unknown input format type.